Solution for CF-718C (#2271)
cpinitiative/usaco-guide: solutions/cf-718C.mdx

## Implementation

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

#define FF first
#define SS second

using ll = long long;
using pll = pair<ll, ll>;
using Matrix = array<array<ll, 2>, 2>;

const ll MOD = 1e9 + 7;

Matrix multiply(const Matrix &a, const Matrix &b) {
	return {(a[0][0] * b[0][0] + a[0][1] * b[1][0]) % MOD, 
		(a[0][0] * b[0][1] + a[0][1] * b[1][1]) % MOD,
		(a[1][0] * b[0][0] + a[1][1] * b[1][0]) % MOD,
		(a[1][0] * b[0][1] + a[1][1] * b[1][1]) % MOD};
}

pll multiply(const pll &a, const Matrix &b) {
	return {(a.FF * b[0][0] + a.SS * b[1][0]) % MOD, 
		(a.FF * b[0][1] + a.SS * b[1][1]) % MOD};
}

Matrix pow(ll n) { 
	Matrix res = {1, 0, 0, 1};
	Matrix base = {0, 1, 1, 1};
	for (; n; n >>= 1) {
		if (n & 1) {
			res = multiply(res, base);
		}
		base = multiply(base, base);
	}
	return res;
}

int N, Q;

> pll tree[400005];
> ll lazy[400005];

Reviewer: ```suggestion
const int MAX = 400005;

pll tree[MAX];
ll lazy[MAX];
```

Author: done

pll merge(const pll &a, const pll &b) {
	return {(a.FF + b.FF) % MOD, (a.SS + b.SS) % MOD};
}

void pushdown(int t) {
	if (lazy[t]) {
		tree[t << 1] = multiply(tree[t << 1], pow(lazy[t]));
		lazy[t << 1] += lazy[t];
		tree[t << 1 | 1] = multiply(tree[t << 1 | 1], pow(lazy[t]));
		lazy[t << 1 | 1] += lazy[t];
		lazy[t] = 0;
	}
}

void update(int l, int r, ll v, int t = 1, int tl = 1, int tr = N) {
	if (r < tl || tr < l) {
		return;
	}
	if (l <= tl && tr <= r) {
		tree[t] = multiply(tree[t], pow(v));
		lazy[t] += v;
		return;
	}
	pushdown(t);

	int tm = (tl + tr) >> 1;
	update(l, r, v, t << 1, tl, tm);
	update(l, r, v, t << 1 | 1, tm + 1, tr);
	tree[t] = merge(tree[t << 1], tree[t << 1 | 1]);
}

ll query(int l, int r, int t = 1, int tl = 1, int tr = N) {
	if (r < tl || tr < l) {
		return 0;
	}
	if (l <= tl && tr <= r) {
		return tree[t].FF;
	}
	pushdown(t);

	int tm = (tl + tr) >> 1;
	return (query(l, r, t << 1, tl, tm) + query(l, r, t << 1 | 1, tm + 1, tr)) % MOD;
}

int main() {
	cin >> N >> Q;

> 	fill_n(tree, 400005, pll{0, 1});

Reviewer: ```suggestion
	fill_n(tree, MAX, pll{0, 1});
```

$$

In that way, we can utilize lazy propgation on the segment tree. We will store an integer tag in each node denoting the lazy update, which is denotes that we must cycle every node in its subtre by the value of the tag. When propogating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 

## Implementation

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

#define FF first
#define SS second

using ll = long long;
using pll = pair<ll, ll>;
using Matrix = array<array<ll, 2>, 2>;

const int MAXN = 1e5 + 5;
const ll MOD = 1e9 + 7;

Matrix multiply(const Matrix &a, const Matrix &b) {
	return {(a[0][0] * b[0][0] + a[0][1] * b[1][0]) % MOD, 
		(a[0][0] * b[0][1] + a[0][1] * b[1][1]) % MOD,
		(a[1][0] * b[0][0] + a[1][1] * b[1][0]) % MOD,
		(a[1][0] * b[0][1] + a[1][1] * b[1][1]) % MOD};
}

pll multiply(const pll &a, const Matrix &b) {
	return {(a.FF * b[0][0] + a.SS * b[1][0]) % MOD, 
		(a.FF * b[0][1] + a.SS * b[1][1]) % MOD};
}

Matrix pow(ll n) { 
	Matrix res = {1, 0, 0, 1};
	Matrix base = {0, 1, 1, 1};

> 	for (; n; n >>= 1) {

Reviewer: explain this

$$

We can use binary exponentiation to quickly compute powers of the two-by-two matrix. 

Secondly, we need to utilize the distribute property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:

$$
(m_1 + m_2 + \dots + m_n) \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} = m_1 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + m_2 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + \dots + m_n \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
$$

In that way, we can utilize lazy propgation on the segment tree. We will store an integer tag in each node denoting the lazy update, which is denotes that we must cycle every node in its subtre by the value of the tag. When propogating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 

## Implementation

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

#define FF first

> #define SS second

Reviewer 2: use only the crap here
https://github.com/cpinitiative/usaco-guide/issues/806

Author: well good thing this isn't bronze through gold then

Reviewer 2: I hate you so much.
![image](https://user-images.githubusercontent.com/55369003/143468481-d272024c-072b-4791-b033-360ff2b59408.png)

\lh

using ll = long long;
using pll = pair<ll, ll>;
using Matrix = array<array<ll, 2>, 2>;

> const int MAXN = 1e5 + 5;

Reviewer 2: replace w/ 1e5 + 1
```suggestion
const int MAXN = 1e5 + 1;
```

const ll MOD = 1e9 + 7;

> // multiplies two two-by-two matricies

Reviewer 2: wrap matrix functions in a codesnip

Matrix multiply(const Matrix &a, const Matrix &b) {
	return {(a[0][0] * b[0][0] + a[0][1] * b[1][0]) % MOD, 
		(a[0][0] * b[0][1] + a[0][1] * b[1][1]) % MOD,
		(a[1][0] * b[0][0] + a[1][1] * b[1][0]) % MOD,
		(a[1][0] * b[0][1] + a[1][1] * b[1][1]) % MOD};
}

// multiplies a one-by-two matrix with a two-by-two matrix
pll multiply(const pll &a, const Matrix &b) {
	return {(a.FF * b[0][0] + a.SS * b[1][0]) % MOD, 
		(a.FF * b[0][1] + a.SS * b[1][1]) % MOD};
}

// returns {0, 1, 1, 1,} to the power of n
Matrix pow(ll n) { 
	Matrix res = {1, 0, 0, 1};  // identity matrix
	Matrix base = {0, 1, 1, 1};  // fibonacci matrix

> 	for (; n; n >>= 1) {. // binary exponentiation (log n)

Reviewer 2: dustin did you mean to put a period here

Reviewer 2: uuuhhh

Reviewer 3: ^^

	if (lazy[t] == 0)
		return;
	tree[t << 1] = multiply(tree[t << 1], pow(lazy[t]));
	lazy[t << 1] += lazy[t];
	tree[t << 1 | 1] = multiply(tree[t << 1 | 1], pow(lazy[t]));
	lazy[t << 1 | 1] += lazy[t];
	lazy[t] = 0;
}

// cycle range from l to r by v
void update(int l, int r, ll v, int t = 1, int tl = 1, int tr = N) {
	if (r < tl || tr < l) {
		return;
	}
	if (l <= tl && tr <= r) {
		tree[t] = multiply(tree[t], pow(v));
		lazy[t] += v;
		return;
	}
	pushdown(t);

	int tm = (tl + tr) >> 1;
	update(l, r, v, t << 1, tl, tm);
	update(l, r, v, t << 1 | 1, tm + 1, tr);
	tree[t] = merge(tree[t << 1], tree[t << 1 | 1]);
}

// query sum from l to r
ll query(int l, int r, int t = 1, int tl = 1, int tr = N) {
	if (r < tl || tr < l) {
		return 0;
	}
	if (l <= tl && tr <= r) {
		return tree[t].FF;
	}
	pushdown(t);

	int tm = (tl + tr) >> 1;
	return (query(l, r, t << 1, tl, tm) + query(l, r, t << 1 | 1, tm + 1, tr)) % MOD;
}

int main() {
	cin >> N >> Q;
	fill_n(tree, MAXN * 4, pll{0, 1});
	for (int i = 1; i <= N; i++) {
		ll a; 
		cin >> a;
		update(i, i, a);
	}

	while (Q--) {
		int t; 
		cin >> t;
		if (t == 1) {
			int l, r; 
			ll v;
			cin >> l >> r >> v;
			update(l, r, v);

> 		}
> 		else if (t == 2) {

Reviewer 2: ```suggestion
		} else if (t == 2) {
```

---
id: cf-718C

> source: cf

Reviewer 3: ```suggestion
source: CF
```

title: Sasha and Array
author: Dustin Miao
---

## Explanation

Firstly, a note on notation: let $F_i$ denote the $i$th fibonacci number. That is,

$$
F_i = 
\begin{cases} 
	0 & i = 0 \\ 
	1 & i = 1 \\

> 	F_{i - 1} + F{i - 2} & i \geq 2 \\

Reviewer 3: ```suggestion
	F_{i - 1} + F_{i - 2} & i \geq 2 \\
```

We will use the term _cycle by k_ to denote transforming some pair $(F_{i - 1}, F_i)$ to $(F_{i - 1 + k}, F_{i + k})$. By default, the term _cycle_ refers to _cycle by 1_.

For each update, we need to cycle each leaf node by some amount $x$ such that the value in a node affected goes from $(F_{v - 1}, F_v)$ to $(F_{v - 1 + x}, F_{v + x})$. We can do this using matrix exponentiation.

First note that for a matrix $\begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}$, we can cycle it by multiplying by $\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}$. That is,

$$
\begin{pmatrix} F_i & F_{i + 1} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
$$

Inductively, we can cycle by $k$ by multiplying multiple times:

$$
\begin{pmatrix} F_{i - 1 + k} & F_{i + k} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}^k
$$

We can use binary exponentiation to quickly compute powers of the two-by-two matrix. 

> Secondly, we need to utilize the distribute property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:

Reviewer 3: ```suggestion
Secondly, we need to utilize the distributive property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
```

Reviewer 3: ```suggestion
Secondly, we need to utilize the distribute property of matrix multiplication to square matrices. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
```

$$
(m_1 + m_2 + \dots + m_n) \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} = m_1 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + m_2 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + \dots + m_n \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
$$

> In that way, we can utilize lazy propgation on the segment tree. We will store an integer tag in each node denoting the lazy update, which is denotes that we must cycle every node in its subtre by the value of the tag. When propogating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 

Reviewer 3: ```suggestion
In that way, we can utilize lazy propagation on the segment tree. We will store an integer tag in each node denoting the lazy update, which denotes that we must cycle every node in its subtree by the value of the tag. When propagating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 
```
